modules/demo/ssr/point-cloud/plugins/point-cloud/index.js

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) 2021, NVIDIA CORPORATION.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+const wrtc            = require('wrtc');
+const {RenderCluster} = require('../../render/cluster');
+
+const {create: shmCreate, detach: shmDetach} = require('shm-typed-array');
+
+module.exports         = graphSSRClients;
+module.exports.graphs  = Symbol('graphs');
+module.exports.clients = Symbol('clients');
+
+/**
+ *
+ * @param {import('fastify').FastifyInstance} fastify
+ */
+function graphSSRClients(fastify) {
+  const graphs  = Object.create(null);
+  const clients = Object.create(null);
+
+  fastify.decorate(module.exports.graphs, graphs);
+  fastify.decorate(module.exports.clients, clients);
+
+  setInterval(layoutAndRenderGraphs(clients));
+
+  return {onConnect, onData, onClose, onError: onClose};
+
+  async function onConnect(sock, peer) {
+    const {
+      width      = 800,
+      height     = 600,
+      layout     = true,
+      g: graphId = 'default',
+    } = sock?.handshake?.query || {};
+
+    const stream = new wrtc.MediaStream({id: `${sock.id}:video`});
+    const source = new wrtc.nonstandard.RTCVideoSource({});
+
+    clients[stream.id] = {
+      video: source,
+      state: {},
+      event: {},
+      props: {width, height, layout},
+      frame: shmCreate(width * height * 3 / 2),
+      peer: peer,
+    };
+
+    stream.addTrack(source.createTrack());
+    peer.streams.push(stream);
+    peer.addStream(stream);
+  }
+
+  function onData(sock, peer, message) {
+    const [stream] = peer?.streams || [];
+    if (stream && !peer.destroyed && !peer.destroying) {
+      const {type, data} = (() => {
+        try {
+          return JSON.parse('' + message);
+        } catch (e) { return {}; }
+      })();
+      switch (data && type) {
+        case 'event': {
+          clients[stream.id].event[data.type] = data;
+          break;
+        }
+      }
+    }
+  }
+
+  function onClose(sock, peer) {
+    const [stream] = peer?.streams || [];
+    if (stream) { delete clients[stream.id]; }
+    const {g: graphId = 'default'} = sock?.handshake?.query || {};
+    if (graphId in graphs) {
+      if ((graphs[graphId].refCount -= 1) === 0) {  //
+        delete graphs[graphId];
+      }
+    }
+  }
+}
+
+function layoutAndRenderGraphs(clients) {
```

Comment:
Could be renamed to PointCloud instead? 
```suggestion
function layoutAndRenderPointCloud(clients) {
```